nemo/collections/nlp/models/language_modeling/megatron_bert_model.py

```diff
@@ -152,26 +237,56 @@ def training_step(self, batch, batch_idx):
             self._reduced_sop_loss_buffer = []
-        return loss
 
-    def validation_step(self, batch, batch_idx):
-        tokens, types, sentence_order, loss_mask, lm_labels, padding_mask = self.process_batch(batch)
-        if not self.cfg.bert_binary_head:
-            types = None
-        output_tensor = self(tokens, padding_mask, token_type_ids=types, lm_labels=lm_labels)
-        loss_dict = self.loss_func(loss_mask, sentence_order, output_tensor)
-        if 'sop loss' in loss_dict:
-            lm_loss = loss_dict['lm loss']
-            sop_loss = loss_dict['sop loss']
-            loss = lm_loss + sop_loss
+        return loss_mean[0]
+
+    def _append_module_grads(self, module, grads):
+        for param in module.parameters():
+            if getattr(param, 'sequence_parallel_enabled', False):
+                if self.megatron_amp_o2:
+                    grad = param.main_grad
+                else:
+                    grad = param.grad
+                grads.append(grad.data)
+
+    def allreduce_sequence_parallel_gradients(self):
+        """ All-reduce layernorm parameters across model parallel nodes when sequence parallelism is used.
+            Modified from megatron-lm:
+            https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/blob/3f91f09bb2ab32f9904b47f46f19d2fc3f518ed8/megatron/training.py#L425
+        """
+
+        grads = []
+        if isinstance(self.model, list):
+            for module in self.model:
+                self._append_module_grads(module, grads)
         else:
-            lm_loss = loss_dict['lm loss']
-            loss = lm_loss
-        reduced_loss = average_losses_across_data_parallel_group([loss])
-        return reduced_loss
+            self._append_module_grads(self.model, grads)
+
+        coalesced = torch._utils._flatten_dense_tensors(grads)
+        torch.distributed.all_reduce(coalesced, group=parallel_state.get_tensor_model_parallel_group())
+        for buf, synced in zip(grads, torch._utils._unflatten_dense_tensors(coalesced, grads)):
+            buf.copy_(synced)
+
+    def validation_step(self, batch, batch_idx):
+        batch_for_pipeline = self.process_batch(batch)
+        tensor_shape = [self.cfg.encoder_seq_length, self.cfg.micro_batch_size, self.cfg.hidden_size]
+
+        losses_reduced_per_micro_batch = forward_backward_no_pipelining(
+            forward_step_func=self.get_forward_output_and_loss_func(),
+            batch=batch_for_pipeline,
+            model=self.model,
+            forward_only=True,
+            tensor_shape=tensor_shape,
+            dtype=self.autocast_dtype,
+        )
+        loss_tensors_list = [loss_reduced['avg'] for loss_reduced in losses_reduced_per_micro_batch]
+        loss_tensor = torch.vstack(loss_tensors_list)
+        loss_mean = loss_tensor.mean(axis=0)
+        return loss_mean[0]
 
     def validation_epoch_end(self, outputs):
         averaged_loss = torch.stack(outputs).mean()
```

Comment:
same comments as training loss average. where is the average across data parallel group happening? @ericharper , I notice there is no explicit average in the GPT model code. Could you confirm?